warehouse/forklift/legacy.py

```diff
@@ -631,6 +674,13 @@ def full_validate(self):
                 "Include at least one message digest."
             )
 
+        # Dynamic is only allowed with metadata version 2.2+
+        if self.dynamic.data and self.metadata_version.data not in ["2.2", "2.3"]:
```

Comment:
```suggestion
        metadata_version = packaging.version.Version(form.metadata_version.data)
        if metadata_version >= packaging.version.Version("2.2")
```

Reply:
Note that this means that we allow 3.0 here (i deleted the test for that from the old PR)